### DISCUSSION TITLE
Font suggestions

### DISCUSSION THREAD

**Discussion by westwick** [sourcelair/xterm.js#5167] [Q&A]
I'm trying to write a telnet client to replace a really old one, but I liked the way it displayed the font and I can't get it to match at all, so was just looking if anyone had any ideas or suggestions for things to try.

It's just the "Terminal" windows system font, size 9 (in the original program, which seems small, but its pretty normal sized really). So I'm not sure if this is some kind of font scaling issue, difference in aspect ratio, or maybe the conversion from .fon to .woff is messing it up. 

Here's the original client I'm trying to match
![image](https://github.com/user-attachments/assets/38171c13-2333-4344-bbc7-949186c6d66f)
![image](https://github.com/user-attachments/assets/98a8e6cd-ab20-42fe-9131-613e907d9bb4)

Thanks for any ideas!!

**Comment by westwick** [comment]
I got a pretty close approximation using "Perfect DOS VGA 437" font (https://www.dafont.com/perfect-dos-vga-437.font)

Had to also disable font-bolding, and set letter-spacing to -1px. It's still very _slightly_ different, but I think I'll call it good enough for now.

Two comparisons:
Original:
![image](https://github.com/user-attachments/assets/b25efe33-2862-4001-bcdf-1ebcaa609e83)

xTerm:
![image](https://github.com/user-attachments/assets/89da3a8c-44f5-4d24-b7e8-b93c8d2e4b74)

Obviously there's a few other small issues (font is slightly bigger, colors are a bit different, I can change those. line height seems very slightly off also), but it looks pretty good overall. You can notice the biggest difference in the font though with how the 'a' and the 'j' are aligned next to each other in the two images. 

Then there's also that issue of the background information disappearing for the usernames, but that's something different, I think I'm losing some ANSI control sequences somewhere, maybe xterm is handling the 'reset' sequence slightly different, not sure.

**Reply by jerch** [reply]
Imho you background issue is also color related. You might want to explicitly set the [color scheme of the windows console](https://en.wikipedia.org/wiki/ANSI_escape_code#Colors) as theme in xtermjs. See [ITerminalOptions.theme](https://xtermjs.org/docs/api/terminal/interfaces/iterminaloptions/#optional-theme) for that.

**Reply by westwick** [reply]
thanks, you are right, I had to explicitly set the 'black' color in theme to #000 in addition to 'background'

**Comment by westwick** [comment]
one other thing I'm curious about too (thought I'd ask here and not pollute the discussion with a bunch of Qs), is how xterm handles the clear screen command \033[2J which on BBS's is usually accompanied by moving the cursor position somewhere.

currently the functionality i get on older telnet clients is that it clears the screen and then starts drawing content from the top of the viewport: 
![image](https://github.com/user-attachments/assets/03021dd2-5044-45e4-b7b3-26b0ebceaf3d)

but with xterm, it clears the screen and then draws new lines/empty divs that end up pushing the content after the clear screen to the botom of the viewport: 
![image](https://github.com/user-attachments/assets/539b0454-95a8-4fd4-a235-607b389a8a55)



which then results in weird drawing issues because it expects that content to have been at the top, and then it starts drawing over it because the cursor position doesnt align that that content that was just drawn.

I searched thru the docs and issues but couldn't find anything related to this, is there any kind of setting for this or do you have any suggestions on how to handle this kind of scenario, or if I had to write a custom solution where I would start?

**Reply by jerch** [reply]
We had a longer discussion about ED behavior here #1727. In short - ED2 is meant to clear the current viewport and nothing more (no scrolling out, no cursor movement), which is what xtermjs currently does.

Imho you are missing the cursor movement to 1;1 here, which can be fixed by a CUP after ED2 ( CUP is `\x1b[H`). So sending `\x1b[2J\x1b[H` should do the trick.
